### ISSUE TITLE
Installation fails on MacOS Monterey, Apple M1 Pro MBP

### ISSUE THREAD

**Issue by ayanb73** [GrossfieldLab/loos#88]
**Describe the bug**
When I attempt to install loos, the build fails partway through. 

**To Reproduce**
- On an appropriate machine: git clone https://github.com/GrossfieldLab/loos.git
- Create conda environment: conda create -n loos
- Activate: conda activate loos
- Navigate to the loos directory: cd loos
- Then inside the loos directory: bash conda_install.sh 

**Expected behavior**
The command {python -c 'import loos'} should be successful with no errors. 

**LOOS version and platform
Which version you're running, on what OS, and how it was built (e.g. conda, system libraries, manually installed dependencies)
The provided conda_install.sh script was utilized using a miniconda installation, on an Apple M1 Pro MacBook Pro (2021) running MacOS Monterey. In addition, a manual build was attempted (on suggestion from @lgsmith. After re-cloning, the following was ran:
> cd loos
> mkdir build
> cd build
> cmake --debug-output --trace-exp .. >& cmake.log
The cmake.log file is attached below. 
[cmake.log](https://github.com/GrossfieldLab/loos/files/9861630/cmake.log)

**Additional context**
The output from the conda_install.sh run is as follows. 
> `make[2]: *** No rule to make target `/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform /Developer/SDKs/MacOSX11.0.sdk/usr/lib/libm.tbd', needed by `src/libloos.dylib'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[ 14%] Building CXX object src/CMakeFiles/loos.dir/scanner.cc.o
/Users/ayan/Documents/loos/src/xtcwriter.cpp:437:29: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
      if (fabs(lf) > INT_MAX-2) 
                   ~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:451:29: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
      if (fabs(lf) > INT_MAX-2)
                   ~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:488:48: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
        (float)maxint[2] - (float)minint[2] >= INT_MAX-2) {
                                            ~~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:487:48: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
        (float)maxint[1] - (float)minint[1] >= INT_MAX-2 ||
                                            ~~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:486:55: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
    if ((float)maxint[0] - (float)minint[0] >= INT_MAX-2 ||
                                            ~~ ~~~~~~~^~
5 warnings generated.
make[1]: *** [src/CMakeFiles/loos.dir/all] Error 2
make: *** [all] Error 2`

**Comment by agrossfield** [comment]
Hi,

I'm sorry you're having trouble. The way you ran things, however, isn't the way INSTALL.md indicates, and I'm pretty sure that's why it's not working for you. 

I suggest you get rid of the loos conda env, and (with just the base conda active) run the command suggested:

```
 ./conda_build.sh -e loos -j 8 -i
```

This will create a new conda env called loos, install the necessary packages, then build and install loos. The fact that it's invoking Xcode stuff makes me think that something didn't get installed correctly -- the whole point of the way we do it is to make sure we use the conda toolchain start to finish. 

I confess I would have expected what you did to work as well, but I've seen conda do weird things before. 

Again, I'm sorry the process isn't working for you.

Alan

PS Alternatively, I expect the conda-forge recipe to be accepted in the next couple of days. After that, you'll be able to just say `conda install -c conda-forge loos` to get everything you need to use existing tools and write new python tools.

**Comment by ayanb73** [comment]
Hi Alan, 

Thanks for your response. I attempted a new installation on a clean environment using the command you provided. I attempted this using both libnetcdf=4.9.0 and libnetcdf=4.8.1, and ran into the same issue both times. The error I receive is the same, with something new added on:
```
make[2]: *** No rule to make target `/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/usr/lib/libm.tbd', needed by `src/libloos.dylib'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[ 14%] Building CXX object src/CMakeFiles/loos.dir/scanner.cc.o
/Users/ayan/Documents/loos/src/xtcwriter.cpp:437:29: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
      if (fabs(lf) > INT_MAX-2) 
                   ~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:451:29: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
      if (fabs(lf) > INT_MAX-2)
                   ~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:488:48: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
        (float)maxint[2] - (float)minint[2] >= INT_MAX-2) {
                                            ~~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:487:48: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
        (float)maxint[1] - (float)minint[1] >= INT_MAX-2 ||
                                            ~~ ~~~~~~~^~
/Users/ayan/Documents/loos/src/xtcwriter.cpp:486:55: warning: implicit conversion from 'int' to 'float' changes value from 2147483645 to 2147483648 [-Wimplicit-const-int-float-conversion]
    if ((float)maxint[0] - (float)minint[0] >= INT_MAX-2 ||
                                            ~~ ~~~~~~~^~
5 warnings generated.
make[1]: *** [src/CMakeFiles/loos.dir/all] Error 2
make: *** [all] Error 2
*** Installing LOOS ***
-- Install configuration: ""
ERROR: Invalid requirement: '/Users/ayan/Documents/loos/build/src/pyloos'
Hint: It looks like a path. File '/Users/ayan/Documents/loos/build/src/pyloos' does not exist.
CMake Error at src/cmake_install.cmake:45 (file):
  file INSTALL cannot find "/Users/ayan/Documents/loos/build/src/_pyloos.so":
  No such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:42 (include)
```
I provide the above in case it's useful, but I'll look forward to the conda-forge recipe!

All the best, 
Ayan

**Comment by agrossfield** [comment]
Also, can you redirect conda_build's output to a file and attach it? I suspect whatever is going wrong is happening long before what we're seeing.

**Comment by ayanb73** [comment]
tried libnetcdf=4.8.0, which did not work. But unfortunately for versions earlier than that I don't think the libnetcdf package is available for an ARM64 architecture on OSX. Attached is an output file of a fresh env, with the build directory removed. 
[conda_build.txt](https://github.com/GrossfieldLab/loos/files/9862698/conda_build.txt)

**Comment by lgsmith** [comment]
I've looked at whether it's possible to downgrade the netcdf version further, and it appears not to be for `osx-arm64`. I suspected that this was what Ayan was running into, but I just checked with conda search and it isn't an option. Here's the filtered output from the `conda search` command I ran to conclude this.
```
conda search -i libnetcdf[subdir=osx-arm64] -c conda-forge | grep -e version
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.0
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.8.1
version     : 4.9.0
version     : 4.9.0
version     : 4.9.0
```
If this is at all relevant, the same command without the `-c conda-forge` optional arg only produces 4.8.0 and 4.8.1.

Are we sure there will be a conda binary for loos on osx-arm64? It seems like it might be hard to make one if the build deps have to come from conda-forge...

**Comment by agrossfield** [comment]
@tromo : do you think we should report the packaging problem to the netcdf conda-forge team? Are we sure this is their problem?

**Comment by agrossfield** [comment]
What confuses me is I'm almost positive there are people who have arm Macs who have loos running, or at least I thought there were. I'll need to get ahold of one to try to figure this out (mine is one of the last generation of intel ones).

**Comment by agrossfield** [comment]
Is that file supplied by cmake or libnetcdf? We need to report the bug to whoever is packaging it

**Comment by tromo** [comment]
It's supplied by libnetcdf.  My "foo" test env doesn't have cmake installed...

**Comment by agrossfield** [comment]
I just confirmed it, and I’m opening an issue on the feedstock repo right now. Reassuring that it’s not our screwup, but still really quite annoying.

________________________________
Dr. Alan Grossfield
Dept of Biochemistry and Biophysics
University of Rochester Medical Center
Phone: 585 276 4193
http://membrane.urmc.rochester.edu
https://orcid.org/0000-0002-5877-2789
Pronouns: He/his



From: Tod Romo ***@***.***>
Reply-To: GrossfieldLab/loos ***@***.***>
Date: Tuesday, October 25, 2022 at 4:17 PM
To: GrossfieldLab/loos ***@***.***>
Cc: "Grossfield, Alan" ***@***.***>, Comment ***@***.***>
Subject: [EXT] Re: [GrossfieldLab/loos] Installation fails on MacOS Monterey, Apple M1 Pro MBP (Issue #88)


It's supplied by libnetcdf. My "foo" test env doesn't have cmake installed...

**Comment by lgsmith** [comment]
On the other hand we did a good deed and can reduce restrictions on our
install, which seems like it'll be helpful.

I'm wondering why 4.8.1 didn't fix the issue, if it doesn't have that cmake
weirdness in it. Maybe the osx-arm64 feedstock got that particular bug
earlier in the version tree?

**Comment by agrossfield** [comment]
Opened https://github.com/conda-forge/libnetcdf-feedstock/issues/155 to notify the libnetcdf folks about the problem

**Comment by agrossfield** [comment]
Per @tromo's workaround above: you'll need to install libxml2 from conda-forge too -- the bug is in part caused because they now have a dependence on libxml2 but forgot to update their conda recipe. I strongly suspect that if you do that, you can just delete the two fields that have the Xcode crap in them and it will work.

In the longer term, the libnetcdf-feedstock maintainers are on the case, and I strongly suspect they'll release new packages soon, at which point we'll update our recipes as well.

**Comment by agrossfield** [comment]
@lgsmith : I suspect the reason it's hitting different versions of libnetcdf on different architectures is the screwup is in the build.sh script in the feedstock: any package built after they introduced the bug will have this problem, independent of the version of libnetcdf itself.

**Comment by agrossfield** [comment]
The libnetcdf folks pushed a "fix" that broke both linux and Mac -- they fixed half the problem by adding libxml2 to their dependencies, but that left the libm problem. I just opened a pull request (https://github.com/conda-forge/libnetcdf-feedstock/pull/158) that should fix the problem.

**Comment by agrossfield** [comment]
My PR was accepted, so hopefully it will percolate through and this will all be resolved shortly

**Comment by lgsmith** [comment]
When you made the fixes did it build for you? I'm assuming yes...

**Comment by agrossfield** [comment]
I didn't really check -- I was editing their build.sh script, specifically modifying a sed command in there that edited some of their cmake files. I restored the sed command that previously fixed the linux build (assuming nothing had changed) and created an equivalent that fixed the OSX problem. All I did was verify that I got a netCDFTargets.cmake file that looked like the ones that worked. The problem is I don't *really* know what that file is supposed to do.

**Comment by agrossfield** [comment]
I just verified that loos builds correctly from conda_build.sh on linux and Mac. You'll need to re-pull, because I removed the lock of libnetcdf=4.8.1 and at this moment only version 4.9.0 works (the back-port to 4.8.1 has been committed but hasn't propagated to the servers yet).

**Comment by agrossfield** [comment]
Current status: there are conda packages for all Macs and linux. However, if you don't have Xcode installed, they might not work -- for some reason, the build on the conda forge server pulls in the Xcode blas and lapack libraries even though when we build locally it correctly grabs the ones from within conda.

I just want people to know we're aware of the problem and working on it (by we, I mean @tromo is working in it).

**Comment by lgsmith** [comment]
Does 4.0.3 fix this?

**Comment by agrossfield** [comment]
Yes, it will. I just didn’t want it closed until the 4.0.3 package was available, which should be later this week.

Dr. Alan Grossfield
Department of Biochemistry and Biophysics
University of Rochester Medical Center

On Nov 14, 2022, at 4:35 PM, lgsmith ***@***.***> wrote:

﻿

Does 4.0.3 fix this?

**Comment by agrossfield** [comment]
4.0.3 is out in both GitHub and conda-forge, so the fix to this problem is available to everyone.
